The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas.

Let's start by importing the libraries we need and reading the dataset into pandas.

In [1]:
import pandas as pd
autos=pd.read_csv(r'https://raw.githubusercontent.com/tamasszuhai/source_files/master/autos.csv',encoding='Latin-1')
#autos.info()
#autos.head()
autos.shape

(50000, 20)

In [60]:
autos_org=pd.read_csv('autos.csv',encoding='Latin-1')
autos_org.shape

(50000, 20)

In [2]:
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [3]:
autos.dtypes.loc[['price','odometer']]
#autos.dtypes.value_counts()
# price  

price       object
odometer    object
dtype: object

In [4]:
autos.dtypes

dateCrawled            object
name                   object
seller                 object
offerType              object
price                  object
abtest                 object
vehicleType            object
yearOfRegistration      int64
gearbox                object
powerPS                 int64
model                  object
odometer               object
monthOfRegistration     int64
fuelType               object
brand                  object
notRepairedDamage      object
dateCreated            object
nrOfPictures            int64
postalCode              int64
lastSeen               object
dtype: object

We can conclude that the values in `price` and `odometer` columns stored as text, but we need to convert them into numeric types. 

Exploring the null values & columns' name <br>
The column names use camelcase instead of Python's preferred snakecase.
Convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
mapping_dict = {'dateCrawled':'date_crawled',
                'name':'name',
                'seller':'seller',
                'price':'price',
                'abtest':'abtest',
                'offerType':'offer_type',
                'vehicleType':'vehicle_type',
                'yearOfRegistration':'registration_year' ,
                'gearbox':'gearbox',
                'powerPS':'power_ps',
                'model':'model',
                'odometer':'odometer_km',
                'monthOfRegistration': 'registration_month',
                'fuelType':'fuel_type',
                'brand':'brand',
                'notRepairedDamage':'unrepaired_damage',
                'dateCreated':'ad_created',
                'nrOfPictures':'pictures',
                'postalCode':'postal_code',
                'lastSeen':'last_seen'    
}

new_columns=autos.columns.map(mapping_dict)
autos.columns=new_columns

Convert `price` and `odometer` columns into numeric types

In [7]:
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

In [8]:
autos["odometer_km"] = (autos["odometer_km"]
                             .str.replace("km","")
                             .str.replace(",","")
                             .astype(int)
                             )
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32

In [9]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null int32
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null int32
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
pictures              50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int32(2), 

In [10]:
autos.seller.value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [11]:
autos.offer_type.value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [12]:
autos.abtest.value_counts()

test       25756
control    24244
Name: abtest, dtype: int64

In [13]:
autos.pictures.value_counts()

0    50000
Name: pictures, dtype: int64

In [14]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null int32
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null int32
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
pictures              50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int32(2), 

In [15]:
#autos.drop(axis=1, labels='pictures', inplace=True)
autos.columns.sort_values()
sorted(autos.columns)

['abtest',
 'ad_created',
 'brand',
 'date_crawled',
 'fuel_type',
 'gearbox',
 'last_seen',
 'model',
 'name',
 'odometer_km',
 'offer_type',
 'pictures',
 'postal_code',
 'power_ps',
 'price',
 'registration_month',
 'registration_year',
 'seller',
 'unrepaired_damage',
 'vehicle_type']

In [16]:
#autos['seller'].value_counts()
#autos['offer_type'].value_counts()
#autos = autos.drop(["seller", "offer_type"], axis=1)
sorted(autos.columns)

['abtest',
 'ad_created',
 'brand',
 'date_crawled',
 'fuel_type',
 'gearbox',
 'last_seen',
 'model',
 'name',
 'odometer_km',
 'offer_type',
 'pictures',
 'postal_code',
 'power_ps',
 'price',
 'registration_month',
 'registration_year',
 'seller',
 'unrepaired_damage',
 'vehicle_type']

## Explore odometer & price column

In [17]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field.  Additionally, there are more high mileage than low mileage vehicles.

In [18]:
print("There are" , len(autos["price"].unique()), 'unique values in the column "price". ')
print('\n',end='')
print(autos["price"].describe())
autos["price"].value_counts().head()
max(autos["price"].unique())

There are 2357 unique values in the column "price". 

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


99999999

Again, the prices in this column seem rounded, however given there are 2357 unique values in the column, that may just be people's tendency to round prices on the site.


There are 1,421 cars listed with \\$0 price - given that this is only 2% of the of the cars, we might consider removing these rows.  The maximum price is one hundred million dollars, which seems a lot, let's look at the highest prices further.

In [19]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [20]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

There are a number of listings with prices below \\$30, including about 1,500 at \\$0.  There are also a small number of listings with very high values, including 14 at around or over \\$1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \\$1.  We will keep the \$1 items, but remove anything above \\$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [21]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

When removing outliers, we can do <span style="background-color: #F59EBF">df[(df["col"] > x ) &
(df["col"] < y )]</span>, 
but it's more readable to use <span style="background-color: #F59EBF">df[df["col"].between(x,y)]</span>


In [22]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64

## Exploring the date columns

In [23]:
sorted(autos.columns)

['abtest',
 'ad_created',
 'brand',
 'date_crawled',
 'fuel_type',
 'gearbox',
 'last_seen',
 'model',
 'name',
 'odometer_km',
 'offer_type',
 'pictures',
 'postal_code',
 'power_ps',
 'price',
 'registration_month',
 'registration_year',
 'seller',
 'unrepaired_damage',
 'vehicle_type']

There are 5 columns that should represent date values.

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In [24]:
autos.loc[0:4,['date_crawled','ad_created','last_seen']]
print(autos['date_crawled'].dtype)
print(autos['ad_created'].dtype)
print(autos['last_seen'].dtype)
print(autos['registration_month'].dtype)
print(autos['registration_year'].dtype)

object
object
object
int64
int64


the first 10 characters represent the day (e.g. `2016-03-05`). 

In [25]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )


2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [26]:
autos.registration_year.describe()

#Another common operation is the use of boolean vectors to filter the data. 
#The operators are: | for or, & for and, and ~ for not. 
#These must be grouped by using parentheses, 
#since by default Python will evaluate an expression such as df.A > 2 & df.B < 3 as df.A > (2 & df.B) < 3, 
#while the desired evaluation order is (df.A > 2) & (df.B < 3).

autos.loc[~(autos['registration_year'].between(1900,2016)),'registration_year'].count() /len(autos['registration_year'])*100

3.8793369710697

In [27]:
(round(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values(ascending=False),3)
        )

2016-04-03    0.039
2016-03-20    0.038
2016-03-21    0.037
2016-03-12    0.037
2016-03-14    0.037
2016-04-04    0.036
2016-03-07    0.036
2016-04-02    0.035
2016-03-28    0.035
2016-03-19    0.035
2016-03-15    0.034
2016-03-29    0.034
2016-04-01    0.034
2016-03-30    0.034
2016-03-08    0.033
2016-03-09    0.033
2016-03-22    0.033
2016-03-11    0.033
2016-03-23    0.032
2016-03-26    0.032
2016-03-10    0.032
2016-03-31    0.032
2016-03-17    0.032
2016-03-25    0.032
2016-03-27    0.031
2016-03-16    0.030
2016-03-24    0.029
2016-03-05    0.025
2016-03-13    0.016
2016-03-06    0.014
2016-04-05    0.013
2016-03-18    0.013
2016-04-06    0.003
2016-04-07    0.001
Name: date_crawled, dtype: float64

In [28]:
#https://towardsdatascience.com/10-python-pandas-tricks-that-make-your-work-more-efficient-2e8e483808ba


def rule(x, y):
    if x == 'high' and y > 10:
         return 1
    else:
         return 0
df = pd.DataFrame({ 'c1':[ 'high' ,'high', 'low', 'low'], 'c2': [0, 23, 17, 4]})
df['new'] = df.apply(lambda x: rule(x['c1'], x['c2']), axis =  1)
df.head()

c1  c2  new
0  high   0    0
1  high  23    1
2   low  17    0
3   low   4    0

In [29]:
print(autos["ad_created"].str[:10].unique().shape)
autos_ads_cr=(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
)
autos_ads_cr[autos_ads_cr>0.02]
print(autos_ads_cr[autos_ads_cr>0.01])

(76,)
2016-03-05    0.022897
2016-03-06    0.015320
2016-03-07    0.034737
2016-03-08    0.033316
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21    0.037579
2016-03-22    0.032801
2016-03-23    0.032060
2016-03-24    0.029280
2016-03-25    0.031751
2016-03-26    0.032266
2016-03-27    0.030989
2016-03-28    0.034984
2016-03-29    0.034037
2016-03-30    0.033501
2016-03-31    0.031875
2016-04-01    0.033687
2016-04-02    0.035149
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
Name: ad_created, dtype: float64


In [30]:
autos_brand_frq_pct=autos["brand"].value_counts(normalize=True)*100
print(autos_brand_frq_pct[autos_brand_frq_pct.values>5],)
print('\n', end='')
print('The auto brands that represented above the 5% in the population are made up ', \
      round(autos_brand_frq_pct[autos_brand_frq_pct.values>5].sum(),2) , '% of the total listings.', sep='')

volkswagen       21.282817
opel             10.865850
bmw              10.859673
mercedes_benz     9.578915
audi              8.582312
ford              6.963863
Name: brand, dtype: float64

The auto brands that represented above the 5% in the population are made up 68.13% of the total listings.


In [130]:
drop_brands=autos_brand_frq_pct[autos_brand_frq_pct<5].index
keep_brands=autos_brand_frq_pct[autos_brand_frq_pct>5].index
#for i in autos["brand"].unique():
#    if i not in keep_brands:
#        drop_brands.append(i)
#dr_list=list(drop_brands.index)
#drop_brands=autos.drop(autos[autos['brand']==drop_brands].index)

keep_brands=list(keep_brands)
#list_of_values = [3,6]
#y = df[df['A'] in list_of_values]
keep_brands


['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford']

We have filtered for the brands representing the 68.13% of the total listings and made a new df named `keep_autos`.

In [ ]:
keep_autos=autos[autos['brand'].isin(keep_brands)]
keep_autos["brand"].value_counts(normalize=True)*100
#https://stackoverflow.com/questions/12096252/use-a-list-of-values-to-select-rows-from-a-pandas-dataframe
#https://stackoverflow.com/questions/32980087/pandas-not-in-in-and-between
#https://stackoverflow.com/questions/19960077/how-to-implement-in-and-not-in-for-pandas-dataframe

#h1 = df1[~df1['nat_actn_2_3'].isin(['100','101','102','103','104'])]

In [142]:
brand_mean_prices = {}

for brand in keep_brands:
    brand_only = keep_autos[keep_autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices=pd.Series(brand_mean_prices)
brand_mean_prices.sort_values(ascending=False)

audi             9212
mercedes_benz    8536
bmw              8261
volkswagen       5332
ford             3728
opel             2944
dtype: int64

Audi, Mercedes and BMW leads the pack in terms of avg prices, VW is in minddle, and Ford & Opel is the most inexpensive ones.

In [152]:
df_brand_means=pd.DataFrame(brand_mean_prices, columns=["mean_prices"]).sort_values(by="mean_prices", ascending=False)
df_brand_means

mean_prices
audi                  9212
mercedes_benz         8536
bmw                   8261
volkswagen            5332
ford                  3728
opel                  2944

In [157]:
brand_mean_mileage = {}

for brand in keep_brands:
    brand_only = keep_autos[keep_autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

df_brand_means['mean_mileage']=mean_mileage
df_brand_means

mean_prices  mean_mileage
audi                  9212        129492
mercedes_benz         8536        130796
bmw                   8261        132682
volkswagen            5332        128896
ford                  3728        124349
opel                  2944        129383